File: apps/Samples/BitonicSortSmall/BitonicSortSmall.cpp
// Copyright 1993-2010 NVIDIA Corporation.  All rights reserved.
//
// Please refer to the NVIDIA end user license agreement (EULA) associated
// with this source code for terms and conditions that govern your use of
// this software. Any use, reproduction, disclosure, or distribution of
// this software and related documentation outside the terms of the EULA
// is strictly prohibited.

// Modified for NoCL, November 2022.

#include <NoCL.h>
#include <Rand.h>

// Size of arrays being sorted
#define LOCAL_SIZE_LIMIT 1024

// Sort two key/value pairs
inline void twoSort(unsigned *keyA, unsigned* valA,
                    unsigned *keyB, unsigned* valB, unsigned dir)
{
  if ((*keyA > *keyB) == dir) {
    unsigned t;
    t = *keyA; *keyA = *keyB; *keyB = t;
    t = *valA; *valA = *valB; *valB = t;
  }
  noclConverge();
}

// Monolithic bitonic sort kernel for short arrays fitting into local mem
struct BitonicSortLocal : Kernel {
  unsigned *d_DstKey_arg;
  unsigned *d_DstVal_arg;
  unsigned *d_SrcKey_arg;
  unsigned *d_SrcVal_arg;
  unsigned arrayLength;
  unsigned sortDir;

  void kernel() {
    unsigned* l_key = shared.array<unsigned, LOCAL_SIZE_LIMIT>();
    unsigned* l_val = shared.array<unsigned, LOCAL_SIZE_LIMIT>();

    // Offset to the beginning of subbatch and load data
    unsigned* d_SrcKey =
      d_SrcKey_arg + blockIdx.x * LOCAL_SIZE_LIMIT + threadIdx.x;
    unsigned* d_SrcVal =
      d_SrcVal_arg + blockIdx.x * LOCAL_SIZE_LIMIT + threadIdx.x;
    unsigned* d_DstKey =
      d_DstKey_arg + blockIdx.x * LOCAL_SIZE_LIMIT + threadIdx.x;
    unsigned* d_DstVal =
      d_DstVal_arg + blockIdx.x * LOCAL_SIZE_LIMIT + threadIdx.x;
    l_key[threadIdx.x + 0] = d_SrcKey[0];
    l_val[threadIdx.x + 0] = d_SrcVal[0];
    l_key[threadIdx.x + (LOCAL_SIZE_LIMIT / 2)] =
      d_SrcKey[(LOCAL_SIZE_LIMIT / 2)];
    l_val[threadIdx.x + (LOCAL_SIZE_LIMIT / 2)] =
      d_SrcVal[(LOCAL_SIZE_LIMIT / 2)];

    for (unsigned size = 2; size < arrayLength; size <<= 1) {
      // Bitonic merge
      unsigned dir = ((threadIdx.x & (size / 2)) != 0);
      for (unsigned stride = size / 2; stride > 0; stride >>= 1) {
        __syncthreads();
        unsigned pos = 2 * threadIdx.x - (threadIdx.x & (stride - 1));
        twoSort(
          &l_key[pos + 0], &l_val[pos + 0],
          &l_key[pos + stride], &l_val[pos + stride], dir);
      }
    }

    // dir == sortDir for the last bitonic merge step
    {
      for(unsigned stride = arrayLength / 2; stride > 0; stride >>= 1){
        __syncthreads();
        unsigned pos = 2 * threadIdx.x - (threadIdx.x & (stride - 1));
        twoSort(
          &l_key[pos + 0], &l_val[pos + 0],
          &l_key[pos + stride], &l_val[pos + stride], sortDir);
      }
    }

    __syncthreads();
    d_DstKey[0] = l_key[threadIdx.x + 0];
    d_DstVal[0] = l_val[threadIdx.x + 0];
    d_DstKey[(LOCAL_SIZE_LIMIT / 2)] =
      l_key[threadIdx.x + (LOCAL_SIZE_LIMIT / 2)];
    d_DstVal[(LOCAL_SIZE_LIMIT / 2)] =
      l_val[threadIdx.x + (LOCAL_SIZE_LIMIT / 2)];
  }
};

int main()
{
  // Are we in simulation?
  bool isSim = getchar();

  // Array size and number of arrays for benchmarking
  int N = LOCAL_SIZE_LIMIT;
  int batch = isSim ? 4 : 32;

  // Input and output vectors
  simt_aligned unsigned srcKeys[N*batch], srcVals[N*batch];
  simt_aligned unsigned dstKeys[N*batch], dstVals[N*batch];

  // Initialise inputs
  uint32_t seed = 1;
  for (int i = 0; i < N*batch; i++) {
    srcKeys[i] = rand15(&seed);
    srcVals[i] = rand15(&seed);
  }

  // Instantiate kernel
  BitonicSortLocal k;

  // Use a single block of threads per array
  k.blockDim.x = LOCAL_SIZE_LIMIT / 2;
  k.gridDim.x = batch;

  // Assign parameters
  k.d_DstKey_arg = dstKeys;
  k.d_DstVal_arg = dstVals;
  k.d_SrcKey_arg = srcKeys;
  k.d_SrcVal_arg = srcVals;
  k.arrayLength = N;
  k.sortDir = 1;

  // Invoke kernel
  noclRunKernelAndDumpStats(&k);

  // Check result
  bool ok = true;
  for (int b = 0; b < batch; b++)
    for (int i = 0; i < N-1; i++)
      ok = ok && dstKeys[b*N+i] <= dstKeys[b*N+i+1];

  // Display result
  puts("Self test: ");
  puts(ok ? "PASSED" : "FAILED");
  putchar('\n');

  return 0;
}

File: rust/examples/bitonic_sort_small.rs
// Copyright 1993-2010 NVIDIA Corporation.  All rights reserved.
//
// Please refer to the NVIDIA end user license agreement (EULA) associated
// with this source code for terms and conditions that govern your use of
// this software. Any use, reproduction, disclosure, or distribution of
// this software and related documentation outside the terms of the EULA
// is strictly prohibited.

// Modified for Rust NoCL, November 2023.

#![no_std]
#![no_main]
#![allow(unused)]

// Crates being used
// =================

use riscv_rt::entry;

extern crate nocl;
use nocl::*;
use nocl::rand::*;
use nocl::prims::*;

extern crate alloc;
use alloc::vec;
use alloc::vec::*;
use alloc::boxed::*;

// Benchmark
// =========

const LOCAL_SIZE_LIMIT: usize = 1024 as usize;

struct BitonicSortLocal {
  length       : usize,
  sort_dir     : bool,
  d_srckey_arg : Buffer<u32>,
  d_srcval_arg : Buffer<u32>,
  d_dstkey_arg : Buffer<u32>,
  d_dstval_arg : Buffer<u32>
}

impl Code for BitonicSortLocal {

#[inline(always)]
fn run (my : &My, shared : &mut Scratch, params: &mut BitonicSortLocal) {
  let mut l_key = alloc::<u32>(shared, LOCAL_SIZE_LIMIT);
  let mut l_val = alloc::<u32>(shared, LOCAL_SIZE_LIMIT);

  // Offset to the beginning of subBATCH and load data
  let d_srckey = my.block_idx.x * LOCAL_SIZE_LIMIT + my.thread_idx.x;
  let d_srcval = my.block_idx.x * LOCAL_SIZE_LIMIT + my.thread_idx.x;
  let d_dstkey = my.block_idx.x * LOCAL_SIZE_LIMIT + my.thread_idx.x;
  let d_dstval = my.block_idx.x * LOCAL_SIZE_LIMIT + my.thread_idx.x;
  l_key[my.thread_idx.x + 0] = params.d_srckey_arg[d_srckey+0];
  l_val[my.thread_idx.x + 0] = params.d_srcval_arg[d_srcval+0];
  l_key[my.thread_idx.x + (LOCAL_SIZE_LIMIT / 2)] =
    params.d_srckey_arg[d_srckey + (LOCAL_SIZE_LIMIT / 2)];
  l_val[my.thread_idx.x + (LOCAL_SIZE_LIMIT / 2)] =
    params.d_srcval_arg[d_srcval + (LOCAL_SIZE_LIMIT / 2)];

  let mut size = 2;
  while size < params.length {
    // Bitonic merge
    let dir = (my.thread_idx.x & (size / 2)) != 0;
    let mut stride = size / 2;
    while stride > 0 {
      syncthreads();
      let pos = 2 * my.thread_idx.x - (my.thread_idx.x & (stride - 1));
      let pos_plus = pos+stride;
      if (l_key[pos] > l_key[pos_plus]) == dir {
        l_key.swap(pos, pos_plus);
        l_val.swap(pos, pos_plus);
      }
      nocl_converge();
      stride = stride >> 1
    }
    size = size << 1
  }

  // dir == sortDir for the last bitonic merge step
  {
    let mut stride = params.length / 2;
    while stride > 0 {
      syncthreads();
      let pos = 2 * my.thread_idx.x - (my.thread_idx.x & (stride - 1));
      let pos_plus = pos + stride;
      if (l_key[pos] > l_key[pos_plus]) == params.sort_dir {
        l_key.swap(pos, pos_plus);
        l_val.swap(pos, pos_plus);
      }
      nocl_converge();
      stride = stride >> 1
    }
  }

  syncthreads();
  params.d_dstkey_arg[d_dstkey + 0] = l_key[my.thread_idx.x + 0];
  params.d_dstval_arg[d_dstval + 0] = l_val[my.thread_idx.x + 0];
  params.d_dstkey_arg[d_dstkey + (LOCAL_SIZE_LIMIT / 2)] =
    l_key[my.thread_idx.x + (LOCAL_SIZE_LIMIT / 2)];
  params.d_dstval_arg[d_dstval + (LOCAL_SIZE_LIMIT / 2)] =
    l_val[my.thread_idx.x + (LOCAL_SIZE_LIMIT / 2)];
}

}

#[entry]
fn main() -> ! {
  nocl_init();

  // Vector size for benchmarking
  const N : usize = LOCAL_SIZE_LIMIT;
  #[cfg(not(feature = "large_data_set"))]
  const BATCH : usize = 4;
  #[cfg(feature = "large_data_set")]
  const BATCH : usize = 32;

  // Input and output vectors
  let mut srckeys : Vec<u32> = vec![0; N*BATCH];
  let mut srcvals : Vec<u32> = vec![0; N*BATCH];
  let mut dstkeys : Vec<u32> = vec![0; N*BATCH];
  let mut dstvals : Vec<u32> = vec![0; N*BATCH];

  // Initialise inputs
  let mut seed : u32 = 1;
  for i in 0 .. N*BATCH {
    srckeys[i] = rand15(&mut seed);
    srcvals[i] = rand15(&mut seed)
  }

  // Use a single block of threads
  let dims =
    Dims {
      block_dim: Dim2 { x: LOCAL_SIZE_LIMIT/2,
                        y: 1 },
      grid_dim: Dim2 { x: BATCH, y: 1 }
    };
 
  // Kernel parameters
  let mut params =
    BitonicSortLocal {
      length       : N,
      sort_dir     : true,
      d_srckey_arg : srckeys.into(),
      d_srcval_arg : srcvals.into(),
      d_dstkey_arg : dstkeys.into(),
      d_dstval_arg : dstvals.into()
    };

  // Invoke kernel
  let params = nocl_run_kernel_verbose(dims, params);

  // Check result
  let mut ok = true;
  for b in 0..BATCH {
    for i in 0 .. N-1 {
      ok = ok && params.d_dstkey_arg[b*N+i] <= params.d_dstkey_arg[b*N+i+1]
    }
  }

  // Display result
  putstr("Self test: ");
  putstr(if ok { "PASSED" } else { "FAILED" });
  putchar(b'\n');

  // The end
  end()
}
